### PR TITLE
[chore] Simplify email types

### PR DESCRIPTION
**Pull request by elland**
This factors out one of our email types in favour of a validated email, this means a lot of tests that created arbitrary bytestrings became invalid and were replaced with simple valid emails, this might have caused redundancy in some of the golden tests.

https://wearezeta.atlassian.net/browse/WPB-665

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: libs/wire-api/src/Wire/API/User/Identity.hs [hunk 1]
```diff
@@ -417,15 +308,17 @@ lenientlyParseSAMLNameID (Just txt) = do
...
+emailFromSAML = fromJust . emailAddressText . SAMLEmail.render
 
 -- | FUTUREWORK(fisx): if saml2-web-sso exported the 'NameID' constructor, we could make this
 -- function total without all that praying and hoping.
-emailToSAMLNameID :: (HasCallStack) => Email -> SAML.NameID
-emailToSAMLNameID = fromRight (error "impossible") . SAML.emailNameID . fromEmail
+emailToSAMLNameID :: EmailAddress -> Either String SAML.NameID
+emailToSAMLNameID = SAML.emailNameID . fromEmail
+
+-- emailToSAMLNameID mail = either (Left "not an email") (SAML.emailNameID . fromEmail $ mail)
```
Comment (mdimjasevic) [inline libs/wire-api/src/Wire/API/User/Identity.hs:319]: A leftover?

File: libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/Email_user.hs [hunk 2]
```diff
@@ -17,120 +17,81 @@
...
+  unsafeEmailAddress "some" "example"
 
-testObject_Email_user_20 :: Email
+testObject_Email_user_20 :: EmailAddress
 testObject_Email_user_20 =
-  Email
-    { emailLocal = "/\138192ZF\1003769\1027227",
-      emailDomain = "\187749I\1109889~FN\1016516\ENQ3PH\DC4k\1036543\131674{\1046142\ETBH\1020386\DC2IR"
-    }
+  unsafeEmailAddress "some" "example"
```
Comment (mdimjasevic) [inline libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/Email_user.hs:97]: If all the 20 test values are `unsafeEmailAddress "some" "example"`, then we as well might keep only one of them and delete the 19 JSON golden files.

File: libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/NewUserPublic_user.hs [hunk 3]
```diff
@@ -38,7 +38,7 @@ testObject_NewUserPublic_user_1 =
         { newUserDisplayName =
             Name {fromName = "\\sY4]u\1033976\DLE\1027259\FS\ETX \US\ETB\1066640dw;}\1073386@\184511\r8"},
           newUserUUID = Nothing,
-          newUserIdentity = Just (EmailIdentity (Email {emailLocal = "test", emailDomain = "example.com"})),
+          newUserIdentity = Just (EmailIdentity (unsafeEmailAddress "some" "example")),
```
Comment (mdimjasevic) [inline libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/NewUserPublic_user.hs:41]: Here, I see no value in changing the value. Have you changed this in an automated fashion, hence the change?

**Review by mdimjasevic — COMMENTED** [review]
There's a lot of diff in golden tests. They pretty much all seem to be facing the same issue: a change in the string value where no change was needed, at least that's my hunch (e.g., replacing `Email {emailLocal = "OR\32966c", emailDomain = "\RS\ENQr"}` with `unsafeEmailAddress "some" "example"`. If it's the case that the former is not a valid email address, then I can see how the change is justified, but please confirm.

File: libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/NewUserPublic_user.hs (hunk 3 above)
Comment (elland) [reply libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/NewUserPublic_user.hs:41]: Correct :)

File: libs/wire-api/src/Wire/API/User/EmailAddress.hs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    isValidLoc x =
+      notNull x
+        && notAt x
+        && isValid (fromString (x <> "@mail.com"))
+
+    isValidDom :: String -> Bool
+    isValidDom x =
+      notNull x
+        && notAt x
+        && isValid (fromString ("me@" <> x))
```
Comment (mdimjasevic) [inline libs/wire-api/src/Wire/API/User/EmailAddress.hs:110]: There's a better way to do this. `suchThat` can lead to test failures when failing to generate an instance in retries, when shrinking, etc.

File: services/brig/test/integration/API/UserPendingActivation.hs [hunk 5]
```diff
@@ -213,23 +213,26 @@ randomScimUserWithSubjectAndRichInfo richInfo = do
     ( (Scim.User.empty @SparTag userSchemas ("scimuser_" <> suffix) (ScimUserExtra richInfo))
         { Scim.User.displayName = Just ("ScimUser" <> suffix),
           Scim.User.externalId = Just externalId,
-          Scim.User.emails = emails,
+          -- Scim.User.emails = emailFromEmailAddress <$> emails,
```
Comment (mdimjasevic) [inline services/brig/test/integration/API/UserPendingActivation.hs:216]: Fix this test.

File: services/federator/test/integration/Test/Federator/Util.hs [hunk 6]
```diff
@@ -310,17 +310,18 @@ defCookieLabel = CookieLabel "auth"
...
 
-mkEmailRandomLocalSuffix :: (MonadIO m) => Text -> m Email
+mkEmailRandomLocalSuffix :: (MonadIO m) => Text -> m EmailAddress
 mkEmailRandomLocalSuffix e = do
   uid <- liftIO UUID.nextRandom
-  case parseEmail e of
-    Just (Email loc dom) -> pure $ Email (loc <> "+" <> UUID.toText uid) dom
+  case emailAddressText e of
+    -- TODO:
+    Just mail -> pure $ unsafeEmailAddress ((localPart mail) <> "+" <> UUID.toASCIIBytes uid) (domainPart mail)
```
Comment (mdimjasevic) [inline services/federator/test/integration/Test/Federator/Util.hs:324]: Something left to do here?

File: services/spar/src/Spar/Intra/BrigApp.hs [hunk 7]
```diff
@@ -78,24 +78,24 @@ veidFromUserSSOId = \case
     case urefToEmail uref of
       Nothing -> pure $ UrefOnly uref
       Just email -> pure $ EmailAndUref email uref
+  -- TODO: parse fallback emails?
```
Comment (mdimjasevic) [inline services/spar/src/Spar/Intra/BrigApp.hs:81]: Something to do here?

File: services/spar/src/Spar/Scim/User.hs [hunk 8]
```diff
@@ -442,10 +449,13 @@ logScimUserId = logUser . Scim.id . Scim.thing
...
-veidEmail (ST.EmailAndUref email _) = Just email
-veidEmail (ST.UrefOnly _) = Nothing
-veidEmail (ST.EmailOnly email) = Just email
+veidToEmail :: ST.ValidExternalId -> Maybe EmailAddress
+veidToEmail (ST.EmailAndUref email _) = Just email
+veidToEmail (ST.UrefOnly _) = Nothing
+veidToEmail (ST.EmailOnly email) = Just email
+
+vsUserEmail :: ST.ValidScimUser -> Maybe EmailAddress
+vsUserEmail usr = veidToEmail usr.externalId -- TODO:
```
Comment (mdimjasevic) [inline services/spar/src/Spar/Scim/User.hs:458]: TODOs all over the module...

File: services/spar/src/Spar/Scim/User.hs [hunk 9]
```diff
@@ -32,10 +33,11 @@
 -- | Doing operations with users via SCIM.
 --
 -- Provides a 'Scim.Class.User.UserDB' instance.
+-- Exported functions are used in tests.
```
Comment (mdimjasevic) [inline services/spar/src/Spar/Scim/User.hs:36]: See if you can rewrite the tests using the instance instead of individual functions.

**Review by mdimjasevic — COMMENTED** [review]
Some test value instan

File: libs/wire-api/src/Wire/API/User/EmailAddress.hs (hunk 4 above)
Comment (elland) [reply libs/wire-api/src/Wire/API/User/EmailAddress.hs:110]: We can address it if it becomes an issue.

**Review by mdimjasevic — APPROVED** [review]
Assuming you fix the two remaining TODOs, this is good to go.

In `Spar.Scim.User` you still have this comment:
`TODO(elland): verify with fisx if this still applies.`
And another one:
`TODO: validate fallback emails?`

**Pull request merged**